sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/JavaMethodReflect.scala

```diff
@@ -0,0 +1,153 @@
+    } else if (!children.take(2).forall(e => e.dataType == StringType && e.foldable)) {
+      // The first two arguments must be string type.
+      TypeCheckFailure("first two arguments should be string literals")
+    } else if (method == null) {
+      TypeCheckFailure("cannot find a method that matches the argument types")
+    } else {
+      TypeCheckSuccess
+    }
+  }
+
+  override def deterministic: Boolean = false
+  override def nullable: Boolean = true
+  override val dataType: DataType = StringType
+
+  override def eval(input: InternalRow): Any = {
+    var i = 0
+    while (i < argExprs.length) {
+      buffer(i) = argExprs(i).eval(input).asInstanceOf[Object]
+      // Convert if necessary. Based on the types defined in typeMapping, string is the only
+      // type that needs conversion.
+      if (buffer(i).isInstanceOf[UTF8String]) {
+        buffer(i) = buffer(i).toString
+      }
+      i += 1
+    }
+    UTF8String.fromString(String.valueOf(method.invoke(obj, buffer : _*)))
+  }
+
+  @transient private lazy val argExprs: Array[Expression] = children.drop(2).toArray
+
+  /** Name of the class -- this has to be called after we verify children has at least two exprs. */
+  @transient private lazy val className = children(0).eval(null).asInstanceOf[UTF8String].toString
+
+  /** The reflection method. */
+  @transient lazy val method: Method = {
+    val methodName = children(1).eval(null).asInstanceOf[UTF8String].toString
+    findMethod(className, methodName, argExprs.map(_.dataType)).orNull
+  }
+
+  /** If the class has a no-arg ctor, instantiate the object. Otherwise, obj is null. */
+  @transient private lazy val obj: Object = instantiate(className).orNull.asInstanceOf[Object]
+
+  /** A temporary buffer used to hold intermediate results returned by children. */
+  @transient private lazy val buffer = new Array[Object](argExprs.length)
+}
+
+object JavaMethodReflect {
+  /** Mapping from Spark's type to acceptable JVM types. */
+  val typeMapping = Map[DataType, Seq[Class[_]]](
+    BooleanType -> Seq(classOf[java.lang.Boolean], classOf[Boolean]),
+    ByteType -> Seq(classOf[java.lang.Byte], classOf[Byte]),
+    ShortType -> Seq(classOf[java.lang.Short], classOf[Short]),
+    IntegerType -> Seq(classOf[java.lang.Integer], classOf[Int]),
+    LongType -> Seq(classOf[java.lang.Long], classOf[Long]),
+    FloatType -> Seq(classOf[java.lang.Float], classOf[Float]),
+    DoubleType -> Seq(classOf[java.lang.Double], classOf[Double]),
+    StringType -> Seq(classOf[String])
+  )
+
+  /**
+   * Finds a Java method using reflection that matches the given argument types,
+   * and whose return type is string.
+   *
+   * The types sequence must be the valid types defined in [[typeMapping]].
+   *
+   * This is made public for unit testing.
+   */
+  def findMethod(className: String, methodName: String, argTypes: Seq[DataType]): Option[Method] = {
+    val clazz: Class[_] = Utils.classForName(className)
```

Comment:
This will throws `ClassNotFoundException`. We had better catch that and throw `AnalysisException`.